java-checks/src/main/java/org/sonar/java/checks/StringIndexOfRangesCheck.java

```diff
@@ -0,0 +1,189 @@
+        callAsSingletonSecondaryLocation(methodInvocation), null);
+    } else if (endIdxDelta.isPresent() && endIdxDelta.get() > 0) {
+      reportIssue(endIdxExpr, "End index should be at most the length of the string.",
+        callAsSingletonSecondaryLocation(methodInvocation), null);
+    } else if (beginIdxDelta.isPresent() && endIdxDelta.isPresent() && beginIdxDelta.get() > endIdxDelta.get()) {
+      reportBeginLargerThanEnd(methodInvocation, beginIdxExpr, endIdxExpr);
+    }
+  }
+
+  private static List<JavaFileScannerContext.Location> callAsSingletonSecondaryLocation(MethodInvocationTree methodInvocation) {
+    return List.of(new JavaFileScannerContext.Location("Call", methodInvocation));
+  }
+
+  private void reportBeginLargerThanEnd(MethodInvocationTree methodInvocation, ExpressionTree beginIdxExpr, ExpressionTree endIdxExpr) {
+    reportIssue(endIdxExpr, "Begin index should not be larger than endIndex.",
+      List.of(
+        new JavaFileScannerContext.Location("Call", methodInvocation),
+        new JavaFileScannerContext.Location("Begin index", beginIdxExpr)
+      ), null
+    );
+  }
+
+  /**
+   * @param expr    the expression
+   * @param varName name of the variable
+   * @return an Optional containing the difference delta such that the value of expr is var.length() + delta (delta may be negative),
+   * or an empty Optional if the expression is too complex for delta to be computed
+   */
+  private static Optional<Integer> lengthDelta(ExpressionTree expr, String varName) {
+    if (isCallToLengthOnVariable(expr, varName)) {
+      // 1st pattern: var.length()
+      return Optional.of(0);
+    }
+    if (expr instanceof BinaryExpressionTree binaryExpr) {
+      var isPlus = binaryExpr.kind() == Tree.Kind.PLUS;
+      var isMinus = binaryExpr.kind() == Tree.Kind.MINUS;
+      if (!(isPlus || isMinus)) {
+        return Optional.empty();
+      }
+      var leftCst = binaryExpr.leftOperand().asConstant(Integer.class);
+      var rightCst = binaryExpr.rightOperand().asConstant(Integer.class);
+      if (isCallToLengthOnVariable(binaryExpr.leftOperand(), varName) && rightCst.isPresent()) {
+        // 2nd pattern:  var.length() + cst  or  var.length() - cst
+        return isPlus ? rightCst : rightCst.map(x -> -x);
+      } else if (isPlus && leftCst.isPresent()
+        && isCallToLengthOnVariable(binaryExpr.rightOperand(), varName)) {
+        // 3rd pattern: cst + var.length()
+        return leftCst;
+      }
+    }
+    return Optional.empty();
+  }
+
+  private static boolean isCallToLengthOnVariable(ExpressionTree expr, String varName) {
+    if (expr instanceof MethodInvocationTree mit) {
+      return LENGTH_MATCHERS.matches(mit.methodSymbol())
+        && mit.methodSelect() instanceof MemberSelectExpressionTree memberSelect
+        && memberSelect.expression() instanceof IdentifierTree idTree
+        && idTree.symbol().name().equals(varName);
```

Comment:
Here we are also granted to have a MemberSelectExpressionTree since we are using the method matcher
```suggestion
      return LENGTH_MATCHERS.matches(mit.methodSymbol())
        && ((MemberSelectExpressionTree)mit.methodSelect()).expression() instanceof IdentifierTree idTree
        && idTree.symbol().name().equals(varName);
```